keps/sig-node/2371-cri-pod-container-stats/README.md

```diff
@@ -272,6 +272,27 @@ Thus, this KEP largely the plan described [here](#plan), with some changes:
 
 #### Open Questions
 1. For the newly introduced CRI API fields, should there be Windows and Linux specific fields?
+    For the alpha implementation, [PR #102789](https://github.com/kubernetes/kubernetes/pull/102789) added a Linux specific field `LinuxPodSandboxStats`. It also added Windows Specific field called  `WindowsPodSandboxStats` but left it blank.  
+
+    Since this field has been added it is proposed to use the opportunity to create structs that are Specific to Windows.  The motivation
+    behind this is Windows has differences in Memory stats that are specific to its OS and doesn't (in some cases cannot) fill fields such as `rss_bytes`.
+
+    A challenge with new `WindowsPodSandboxStats` which custom fields is that the current calls to CRI endpoint `ListContainerStats`
+    use the exiting `ContainerStats` object which is not Windows/Linux specific.  In the case of the CRI call `ListPodSandboxStats`
+    it would currently return existing fields.  There would be a miss match between the structs returned between these calls unless we update. Making changes to `ContainerStats` in backwards compatible way may be difficult.
+
+    Alternatives include:
+    
+    - Rename the field `LinuxPodSandboxStats` to `PodSandboxStats`.  Windows and
```

Comment:
After going through this I am actually leaning towards this being the best approach.  I am not sure how feasible it is to rename fields in CRI.

Reply:
I am open to this approach, though I don't think we should remove it, but rather add a new object, and append that object to the proto definition (and mark the old one as deprecated). That will allow backwards compatibliity

Reply:
I think the approach of having a single `PodSandboxStats` could work and would provide some consistency / simplify logic in the kubelet.

One concern I have is if we do this then some of the fields like `rss_bytes` wouldn't be available for Windows.
As long as things consuming these stats are aware of this maybe that isn't an issue?
If we do go with this approach let's at least make sure the comments in the proto definitions specify which memory fields are expected to be populated for each OS type.

Reply:
Since these stats aren't currently filled out today (as mentioned below) I don't think this will be an issue.

Reply:
I think the code consuming the metrics (kubelet, primarily) should be able to handle missing metrics. There are fields that have times when they're expected to be not specified

Reply:
I wanted to be able to use this cri only fields KEP to use specific windows stats but based on the research it seems we could fill in many of the missing stats now (and mark things like `rss_bytes` as not used by windows in the comments).  

Making windows specific fields might be something we could look at into the future but feels like it would need its own KEP and specific use cases.  Right now, the consumer of the summary API expects these specific fields and making Windows specific fields creates unnecessary complexity that isn't going to be consumed immediately.

if we have general agreement here, I can update the KEP to reflect this approach of using  a single `PodSandboxStats`

Reply:
Thanks for looking into this, I agree making use of a single struct `PodSandboxStats` just like existing ContainerStats sounds like the most reasonable approach.

Reply:
I think the issue with just mirroring is that now we have a Windows specific struct with Linux fields in it (`MemoryUsage` has `rss_bytes`) so we don't really gain anything and still have to do additional mapping in the kubelet to handle it.

If we had uses cases for extra fields I think it might make sense to do the work outlined in this change to KEP but at this time we don't have anything concrete.  If we went the route of using a generic `PodSandboxStats` in the future we could still add Linux and Windows sub fields for if there was need for specific fields.  Could be something like @haircommander suggested in https://github.com/kubernetes/enhancements/pull/3439#discussion_r927665278

```
message PodSandboxConfig {
    PodSandboxMetadata metadata = 1;
    string hostname = 2;
    string log_directory = 3;
    DNSConfig dns_config = 4;
    map<string, string> labels = 6;
    map<string, string> annotations = 7;

    // Optional configurations specific to Linux hosts. 
    // (currently duplicate of PodSandboxConfig but may be used in future for linux specific stats)
    LinuxPodSandboxConfig linux = 8;
    // Optional configurations specific to Windows hosts.
    PodSandboxConfig podsandboxconfig = 9;

    //Windows Pod SandboxConfig (nothing here now but could be added if we identified usecases in the future 
    // may require a kep )
    message WindowsPodSandboxStats 
}
```

Reply:
@marosset @bobbypage @haircommander I've updated the KEP based on this conversation.  PTAL

Hoping we can get consensus so we can get these changes into CRI for the next contianerd Release.  This would allow us to have Windows support when this KEP goes to Beta.

Reply:
I'm fine with any of the approach discussed here.
I'll also +1 the importance of getting consensus on this soon. We'll want the CRI-API changes to be made the release _before_ this feature is enabled by default in order to give container runtimes adequate time to provide implementations.
Containerd usually doesn't vendor in CRI-API changes until after K8s releases.

Reply:
after much deliberation, we decided to keep the fields separate. I've updated the kep and we can discuss further in a new thread if needed